client/src/Features/MicrosoftLearn/MicrosoftLearnFilterCore.ts

```diff
@@ -0,0 +1,199 @@
+    displayFilters.set(FilterType.Level, [...filteredLevels]);
+    displayFilters.set(FilterType.Type, [...filteredTypes]);
+
+    return displayFilters;
+}        
+
+export const updateURI = (selectedFilters: Map<FilterType, string[]>, searchTerm: string, expandedProducts: string[], productHierarchicalMap: Map<Product, Product[]>) => {
+    let productFilter = selectedFilters.get(FilterType.Product) || [];
+    let roleFilter = selectedFilters.get(FilterType.Role) || [];
+    let typeFilter = selectedFilters.get(FilterType.Type) || [];
+    let levelFilter = selectedFilters.get(FilterType.Level) || [];
+
+    const getProductUri = () => {
+      let parents = [...productHierarchicalMap.keys()]; 
+      let keep: string[] = [];
+      let parentProductFilters: Product[] = parents.filter(p => productFilter.includes(p.id));
+      let invisibleChildren = _.flatten(parentProductFilters.map(parent => productHierarchicalMap.get(parent)?.map(c=>c.id)));
+
+      keep = [...keep, ...productFilter.filter(f => !keep.includes(f) && !invisibleChildren.includes(f))];
+      return 'products='+keep.join('%2C');
+    }
+
+    let productUri = productFilter.length>0? getProductUri() : '';
+    let roleUri = roleFilter.length>0? 'roles=' + roleFilter.join('%2C') : '';
+    let typeUri = typeFilter.length>0? 'types=' + typeFilter.join('%2C') : '';
+    let levelUri = levelFilter.length>0? 'levels=' + levelFilter.join('%2C') : '';
+    let termsUri = searchTerm.length>0? 'terms='+searchTerm : '';
+    let expandedProductsUri = expandedProducts.length>0? 'expanded='+expandedProducts.join('%2C') : '';
+    let finalUri = [productUri, roleUri, levelUri, typeUri, termsUri, expandedProductsUri]
+                    .filter(s=>s.length!==0).join('&')            
+
+    let uri = window.location.protocol + "//" + window.location.host + window.location.pathname + '?' + finalUri;
+    window.history.pushState({path:uri},'',uri);       
+
+    return finalUri;
+}
+
+const getSearchTermFilteredLearnContent = (expressions: RegExp[], content: LearnContent[]): LearnContent[] => {
+    return content
+      .map(course => ({
+        course: course,
+        score: _.sumBy(
+          expressions,
+          singleExpression =>
+            scoreRegex(course.summary, singleExpression) + scoreRegex(course.title, singleExpression, 2)
+        )
+      }))
+      .filter(scouredCourse => scouredCourse.score > 0)
+      .sort((a, b) => b.score - a.score)
+      .map(scoredCourse => scoredCourse.course);
+}
+
+export const loadFiltersFromPath = (catalog: Catalog, productHierarchicalMap: Map<Product, Product[]>) => {
+    let searchParams = window.location.search
+    if(searchParams && searchParams.length>0){
+      let searchParamMap: Map<string, string[]> = new Map<string, string[]>();
+      searchParams.slice(1).split('&').forEach(s=>{
+        let [key, value] = s.split('=');
+        if(value){
+            searchParamMap.set(key, value.split('%2C'));
```

Comment:
**Request:** What should the behavior of code be in case `value == '%2C    %2C %2C...'` (i.e. a combination of '%2C' and blank spaces only)? Should we check for that case as well?